lib/value-pairs/value-pairs.c

```diff
@@ -617,20 +617,26 @@ vp_walker_split_name_to_tokens(vp_walk_state_t *state, const gchar *name)
 
   GPtrArray *array = g_ptr_array_sized_new(VP_STACK_INITIAL_SIZE);
 
-  do
+  while (pos < name_len)
     {
-      pos += strcspn(&name[pos], "@.");
       if (name[pos] == '@')
-        pos = vp_walker_skip_sdata_enterprise_id(name, pos);
-
-      if (name[pos] == '.' || pos == name_len)
+        {
+          pos = vp_walker_skip_sdata_enterprise_id(name, pos);
+        }
+      else if (name[pos] == '.')
         {
           g_ptr_array_add(array, g_strndup(&name[token_start], pos - token_start));
           pos++;
           token_start = pos;
         }
+      else
+        {
+          pos += strcspn(&name[pos], "@.");
+        }
     }
-  while (pos < name_len);
+
+  if ((pos - token_start) > 0)
+    g_ptr_array_add(array, g_strndup(&name[token_start], pos - token_start));
 
```

Comment:
Your solution also sounds more efficient. And from a higher perspective, I
can understand the intention of this code part as well.

On Thu, Mar 17, 2016 at 10:22 AM, László Várady <notifications@github.com>
wrote:

> In lib/value-pairs/value-pairs.c
> <https://github.com/balabit/syslog-ng/pull/969#discussion_r56474953>:
>
> >      }
> > -  while (pos < name_len);
> > +
> > +  if ((pos - token_start) > 0)
> > +    g_ptr_array_add(array, g_strndup(&name[token_start], pos - token_start));
> >
>
> This commit is only for clarification.
>
> if ((pos - token_start) > 0)
>   g_ptr_array_add(array, g_strndup(&name[token_start], pos - token_start));
>
> The additional if statement is needed because @dnsjts
> <https://github.com/dnsjts> didn't want to "duplicate" the loop condition
> (pos < name_len) within the loop (like it was before):
>
> do
>   {
>     /* ... */
>     if (name[pos] == '.' || pos == name_len)
>     /* ... */
>   }while (pos < name_len);
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/balabit/syslog-ng/pull/969/files/d688def497b557392216acfc6425d11ec0468cdc..c4654d2a7adafaca0d4184f2af353fa6df0b0d99#r56474953>
>